hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,224 @@
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      LOG.debug("Performing compaction in non-batched mode, regions={}, batch size={}",
+        regions.size(), regionBatchSize);
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i = 0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for (RegionInfo ri : toCompact) {
+      startCompaction(admin, htd.getTableName(), ri, hcd.getName());
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    int totalCompacted = 0;
+    while (!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri : toCompact) {
+        try {
+          if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+            totalCompacted++;
+            LOG.info("Finished major compaction: table={} region={}, compacted regions={}",
+              htd.getTableName(), ri.getRegionNameAsString(), totalCompacted);
+            compacted.add(ri);
+          }
+        } catch (IOException e) {
+          LOG.warn("Could not get compaction state for region {}", ri.getEncodedName());
+        }
+      }
+      // Update batch: remove compacted regions and add new ones
+      for (RegionInfo ri : compacted) {
+        toCompact.remove(ri);
+        if (regions.size() > 0) {
+          RegionInfo region = regions.remove(0);
+          startCompaction(admin, htd.getTableName(), region, hcd.getName());
+          toCompact.add(region);
+        }
+      }
+      compacted.clear();
+      LOG.debug("Wait for 10 sec, toCompact size={} regions left={} compacted so far={}",
+        toCompact.size(), regions.size(), totalCompacted);
+      Thread.sleep(10000);
+    }
+    LOG.info("Finished major compacting {}. cf={}", htd.getTableName(), hcd.getNameAsString());
+
+  }
+
+  private void startCompaction(Admin admin, TableName table, RegionInfo region, byte[] cf)
+      throws IOException, InterruptedException {
+
+    LOG.info("Started major compaction: table={} region={}", table, region.getRegionNameAsString());
+    admin.majorCompactRegion(region.getRegionName());
+    // Wait until it really starts
+    // but with finite timeout
+    long waitTime = 300000; // 5 min
+    long startTime = EnvironmentEdgeManager.currentTime();
+    while (admin.getCompactionStateForRegion(region.getRegionName()) == CompactionState.NONE) {
+      // Is 1 second too aggressive?
+      Thread.sleep(1000);
+      if (EnvironmentEdgeManager.currentTime() - startTime > waitTime) {
+        LOG.warn("Waited for {} ms to start major compaction on table: {} region: {}. Aborted.",
```

Comment:
instead of "aborted" say precisely what behavior we're changing. (did we abort the compaction? did we abort waiting?)

it looks like we should say something like "continuing as though it started"